### DISCUSSION TITLE
v3.0 deployed on Vercel: There was an error initializing Payload

### DISCUSSION THREAD

**Discussion by wilbertcaba** [payloadcms/payload#6700] [Q&A] [answered]
### Link to reproduction

https://wilbert-caba-portfolio-git-featur-c0d2e2-wilbert-cabas-projects.vercel.app?_vercel_share=XXnAoR3OESpwCi91KlFAQf5QWJnSEMz7

### Payload Version

3.0.0-beta.39

### Node Version

>=18.19.0

### Next.js Version

14.3.0-canary.28

### Describe the Bug

I'm using the API to easily fetch data. It works like a charm locally. I pushed my website to a Github branch which Vercel then deploys for previewing purposes (I haven't merged into the main branch yet). When I preview my deployed branch using Vercel's preview link, I'm encountering the following issues:

## 1. I can't access Payload's admin:
I tried accessing the Payload's admin and I get a 500 error header:

```
Request URL:
https://wilbert-caba-portfolio-git-featur-c0d2e2-wilbert-cabas-projects.vercel.app/admin
Request Method:
GET
Status Code:
500 Internal Server Error
Remote Address:
76.76.21.241:443
Referrer Policy:
strict-origin-when-cross-origin
```

## 2. I can't fetch data
I can't get to display my projects from `/api/Projects` as I get a 500 error when trying to fetch these. I've logged the error in the console and I get "There was an error initializing Payload" from my `try{...}` block:

**projects.tsx**
```javascript
// ...
useEffect(() => {
        const fetchProjects = async () => {
            try {
                const response = await fetch("/api/Projects");
                const data = await response.json();
                console.log("Data received for projects:", data); // On Vercel this returns: "There was an error initializing Payload" 
                setProjects(data.docs);
                setError(null);
            } catch (error) {
                console.log("We got this error: ", error);
                setError("Failed to fetch projects.");
                setProjects([]);
            }
        };
        fetchProjects();
    }, []);
    
// ...
```

Based on what I've read from @jmikrut [on this issue](https://github.com/payloadcms/payload/issues/514): "Vercel is instead a serverless host, which is set up to run serverless functions", Payload couldn't be initialized. Today, Payload v3 supports serverless functions as mentioned [on this blog post ](https://payloadcms.com/blog/30-beta-install-payload-into-any-nextjs-app-with-one-line#blog) by the Payload team: "Regardless, as of this moment, it's all serverless, so you can deploy to Vercel, and it's just going to work right out of the box." so I assume this should work fine, which is not my case.

This is what I'm getting from Vercel's Deployment Details logs:

TL;DR
```

▲ Next.js 14.3.0-canary.28
--
03:35:36.039 |  
03:35:36.126 | Creating an optimized production build ...
03:36:34.912 | ✓ Compiled successfully
03:36:34.914 | Linting and checking validity of types ...
03:36:43.075 | Collecting page data ...
03:36:46.996 | Generating static pages (0/7) ...
03:36:47.781 | Generating static pages (1/7)
03:36:47.813 | Generating static pages (3/7)
03:36:47.813 | Generating static pages (5/7)
03:36:48.047 | ✓ Generating static pages (7/7)
03:36:49.362 | Finalizing page optimization ...
03:36:49.362 | Collecting build traces ...
03:36:57.669 |  
03:36:57.674 | Route (app)                              Size     First Load JS
03:36:57.674 | ┌ ○ /                                    54 kB           159 kB
03:36:57.674 | ├ ○ /_not-found                          904 B          89.4 kB
03:36:57.674 | ├ ƒ /admin/[[...segments]]               615 B           453 kB
03:36:57.674 | ├ ƒ /api/[...slug]                       0 B                0 B
03:36:57.674 | ├ ƒ /api/graphql                         0 B                0 B
03:36:57.674 | ├ ƒ /api/graphql-playground              0 B                0 B
03:36:57.674 | └ ƒ /projects/[slug]                     2.26 kB        96.9 kB
03:36:57.674 | + First Load JS shared by all            88.5 kB
03:36:57.674 | ├ chunks/7023-f16c0603b074b27a.js      31.9 kB
03:36:57.674 | ├ chunks/fd9d1056-271cc4b94c27211e.js  53.5 kB
03:36:57.674 | └ other shared chunks (total)          3.04 kB
03:36:57.674 |  
03:36:57.674 |  
03:36:57.674 | ○  (Static)   prerendered as static content
03:36:57.674 | ƒ  (Dynamic)  server-rendered on demand
03:36:57.674 |  
03:36:57.777 | Done in 82.68s.
03:36:57.859 | Traced Next.js server files in: 33.483ms
03:37:04.861 | Created all serverless functions in: 7.002s
03:37:04.872 | Collected static files (public/, static/, .next/static): 7.59ms
03:37:04.910 | Build Completed in /vercel/output [2m]
03:37:05.100 | Deploying outputs...
03:37:05.101 | Injected preview comments into the build output, you can disable this in your project settings.
03:37:15.089 |  
03:37:15.350 | Deployment completed
03:37:29.700 | Uploading build cache [194.58 MB]...
03:37:34.110 | Build cache uploaded: 4.409s
```

What I can see on these logs is that these `ƒ /api/[...slug]`, `ƒ /api/graphql` and `ƒ /api/graphql-playground` produced 0 bytes.

What's confusing to me, and to my knowledge's extent is somewhat a discrepancy due to my humble ignorance, is that when I scroll to the Functions tab within Deployment Details, I see that these were created? 

<img width="1225" alt="Screenshot 2024-06-08 at 1 19 03 PM" src="https://github.com/payloadcms/payload/assets/36094085/d4719b38-334d-491f-b56d-f711eebb2579">
_(Help wanted trying to understand why the logs show 0 bytes while these Functions tab above shows data)_

Please let me know if I am overlooking something.

### Reproduction Steps

1. Use Payload's API to fetch.
3. Push to a Github project connected to Vercel.
4. Preview site using Vercel's preview link and try:
- accessing the admin
- run a page with dynamic data.

### Adapters and Plugins

postgresAdapter (using Supabase) and s3Storage

**Comment by DanRibbens** [comment, marked as answer]
Are you able to access the admin UI on vercel? Using the url you provided, I'm not able to get the admin to load at all: https://wilbert-caba-portfolio-git-featur-c0d2e2-wilbert-cabas-projects.vercel.app/admin which tells me you have other issues.

Aside from that, you should be able to use a server component and avoid using `fetch`. In a client component, using fetch should be fine but you need to add `'Content-Type': 'application/json'` headers to your fetch request.

Please verify that everything else is working besides your client component issue.

**Reply by wilbertcaba** [reply]
@DanRibbens I ended up setting a Postgres database using Supabase and fetching data on a server component instead of using fetch on the client as per your suggestion. That seemed to do the work, so I truly appreciate your recommendation.

**Comment by jmikrut** [comment]
Hey @wilbertcaba — to build on what Dan said, you should be seeing some errors in your logs on Vercel. Payload does indeed work on Vercel with 3.0, but when I load your website, I see that it produces a 500 error on the client side. Hunt down that error and try to determine what is going on. Is it your database connection string?

Also, the first load JS being zero bytes is expected, because those are API route handlers, and by their nature, API routes do not download any first load JS. 

I will convert this to a discussion but we can continue to try and help here!

**Reply by wilbertcaba** [reply]
@jmikrut got it, now it's clear to me. Thanks a lot for your support, man. Keep up the good work... love where things are going with Payload v3.

**Comment by wilbertcaba** [comment]
Refer to [answer above](https://github.com/payloadcms/payload/discussions/6700#discussioncomment-9727147)

**Comment by ashayramolia** [comment]
Having similar issues, 
Hosting on vercel, Supabase as the Database provider to payload, 
Getting the 500: Internal Server Error.
